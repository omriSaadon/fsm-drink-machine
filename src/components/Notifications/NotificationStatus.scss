.notification {
  width: calc(100% - 10px);
  margin-bottom: 20px;
  height: 40px;
}