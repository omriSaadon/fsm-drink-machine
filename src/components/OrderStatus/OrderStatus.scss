.cancelPayment {
  margin-top: 15px;
}