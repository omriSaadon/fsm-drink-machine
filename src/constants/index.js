import React from 'react';

export const INITIAL_BUDGET = 10;

export const GENERAL_ACTIONS = {
  ADD_BUDGET: 'addBudget',
  ADD_MESSAGE: 'addMessage',
};

export const STATES = {
  IDLE: 'idle',
  ORDERING: 'ordering',
  DELIVERY: 'delivery',
};

export const ACTION_STATES = {
  PAYMENT: 'payment',
  CANCEL: 'cancel',
  PURCHASE: 'purchase',
  DONE: 'done',
};

export const COINS = [
  { key: 'one', label: '1 ₪', value: 1 },
  { key: 'two', label: '2 ₪', value: 2 },
  { key: 'five', label: '5 ₪', value: 5 },
  { key: 'ten', label: '10 ₪', value: 10 },
];

export const DRINKS = [
  {
    key: 'Coke', label: 'Coke', price: 8, icon: <img alt="" src={`${process.env.PUBLIC_URL}/assets/images/coca-cola.jpeg`} />,
  },
  {
    key: 'Sparkling_Water', label: 'Sparkling Water', price: 6, icon: <img alt="" src={`${process.env.PUBLIC_URL}/assets/images/sparkling-water.jpeg`} />,
  },
  {
    key: 'Water', label: 'Water', price: 5, icon: <img alt="" src={`${process.env.PUBLIC_URL}/assets/images/water.jpeg`} />,
  },
  {
    key: 'Ice_Tea', label: 'Ice Tea', price: 7, icon: <img alt="" src={`${process.env.PUBLIC_URL}/assets/images/ice-tea.jpeg`} />,
  },
];

export const FLOW_INITIAL_ELEMENTS = [
  {
    id: STATES.IDLE,
    data: { label: 'Idle' },
    position: { x: 10, y: 100 },
    targetPosition: 'right',
    sourcePosition: 'bottom',
  },
  {
    id: 'payment_needed', data: { label: 'Payment Is Needed' }, position: { x: 300, y: 10 }, targetPosition: 'left', sourcePosition: 'left',
  },
  {
    id: STATES.ORDERING, data: { label: 'Ordering' }, position: { x: 420, y: 240 }, sourcePosition: 'left',
  },
  {
    id: STATES.DELIVERY, data: { label: 'delivery' }, position: { x: 550, y: 100 }, sourcePosition: 'left', targetPosition: 'bottom',
  },
  {
    id: ACTION_STATES.CANCEL, data: { label: 'Cancel Order' }, position: { x: 100, y: 200 }, targetPosition: 'right', sourcePosition: 'top',
  },
  {
    id: 'idle-payment_needed',
    source: STATES.IDLE,
    target: 'payment_needed',
    labelStyle: { fill: 'red', fontWeight: 700 },
    label: 'Clicked item without payment',
  },
  {
    id: 'idle-ordering', source: STATES.IDLE, target: STATES.ORDERING, animated: true, label: 'Payment Succeed', arrowHeadType: 'arrowclosed',
  },
  {
    id: 'ordering-delivery', source: STATES.ORDERING, target: STATES.DELIVERY, animated: true, label: 'Item Selected', arrowHeadType: 'arrowclosed',
  },
  {
    id: 'ordering-cancel',
    source: STATES.ORDERING,
    target: ACTION_STATES.CANCEL,
    label: 'Canceled',
    labelStyle: { fill: 'red', fontWeight: 700 },
    arrowHeadType: 'arrowclosed',
  },
  {
    id: 'cancel-idle', source: ACTION_STATES.CANCEL, target: STATES.IDLE, arrowHeadType: 'arrowclosed',
  },
  {
    id: 'delivery-idle',
    source: STATES.DELIVERY,
    target: STATES.IDLE,
    arrowHeadType: 'arrowclosed',
    animated: true,
    labelBgPadding: [8, 4],
    labelBgBorderRadius: 4,
    labelBgStyle: {
      fill: '#FFCC00', fillOpacity: 1,
    },
    label: 'Finished',
  },
];
