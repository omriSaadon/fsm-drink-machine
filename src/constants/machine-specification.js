import {
  makePayment, addBudget, cancelPayment, makePurchase, addMessage,
} from 'store/fsm_reducer';
import { INITIAL_BUDGET, STATES, ACTION_STATES } from '.';

export default {
  initialState: {
    state: STATES.IDLE,
    context: {
      currentBudget: INITIAL_BUDGET,
      paid: 0,
      message: '',
    },
  },
  generalActions: {
    addBudget,
    addMessage,
  },
  states: {
    [STATES.IDLE]: {
      [ACTION_STATES.PAYMENT]: {
        next: STATES.ORDERING,
        action: makePayment,
      },
    },
    [STATES.ORDERING]: {
      [ACTION_STATES.CANCEL]: {
        next: STATES.IDLE,
        action: cancelPayment,
      },
      [ACTION_STATES.PURCHASE]: {
        next: STATES.DELIVERY,
        action: makePurchase,
      },
      [ACTION_STATES.PAYMENT]: {
        next: STATES.ORDERING,
        action: makePayment,
      },
    },
    [STATES.DELIVERY]: {
      [ACTION_STATES.DONE]: {
        next: STATES.IDLE,
      },
      [ACTION_STATES.PAYMENT]: {
        next: STATES.ORDERING,
        action: makePayment,
      },
    },
  },
};
