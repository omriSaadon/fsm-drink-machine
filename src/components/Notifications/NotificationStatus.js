import React, { useEffect } from 'react';
import { Alert } from 'antd';
import { useStore } from 'store/FsmStore';
import { getMessageType, getMessage } from 'store/fsm_reducer';
import { GENERAL_ACTIONS } from 'constants/index';
import classes from './NotificationStatus.scss';

const NotificationStatus = () => {
  const [state, dispatch] = useStore();

  const message = getMessage(state);
  const messageType = getMessageType(state);

  const onClose = () => {
    dispatch({ generalAction: GENERAL_ACTIONS.ADD_MESSAGE, payload: { message: '', type: null } });
  };

  useEffect(() => {
    const timer = setTimeout(() => {
      onClose();
    }, 5000);
    return () => clearTimeout(timer);
  }, [message]);

  return (
    <div className={classes.notification}>
      {message
            && (
            <Alert
              message={message}
              type={messageType}
              closable
              onClose={onClose}
            />
            )}
    </div>

  );
};

export default NotificationStatus;
