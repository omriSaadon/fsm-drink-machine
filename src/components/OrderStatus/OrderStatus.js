import React from 'react';
import { Card, Button } from 'antd';
import { useStore } from 'store/FsmStore';
import { getPaid, getMachineStatus } from 'store/fsm_reducer';
import { ACTION_STATES, STATES, GENERAL_ACTIONS } from 'constants/index';
import classes from './OrderStatus.scss';

const OrderStatus = () => {
  const [state, dispatch] = useStore();

  const paid = getPaid(state);
  const drinkMachineStatus = getMachineStatus(state);

  const paidText = `You have paid ${paid} â‚ª`;

  const onCancelPament = () => {
    // Cancel payment can be made only on Ordering state, otherwise an error wil appear.
    if (drinkMachineStatus === STATES.ORDERING) {
      dispatch({ state: ACTION_STATES.CANCEL });
    } else {
      dispatch({ generalAction: GENERAL_ACTIONS.ADD_MESSAGE, payload: { message: 'Cancel can be made only on ordering status', type: 'error' } });
    }
  };

  return (
    <Card title="Order Status">
      <div>{paidText}</div>
      <Button className={classes.cancelPayment} onClick={onCancelPament} shape="round" size="small">Cancel Payment</Button>
    </Card>
  );
};

export default OrderStatus;
