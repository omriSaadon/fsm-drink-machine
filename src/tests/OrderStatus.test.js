import React from 'react';
import { screen, fireEvent } from '@testing-library/react';
import { STATES } from 'constants/index';
import OrderStatus from 'components/OrderStatus/OrderStatus';
import NotificationStatus from 'components/Notifications/NotificationStatus';
import { machineRender } from './helpers/machineHelper';
import mockStore from './mocks/mock_context';

test('test on cancel payment in delivery mode - should be blocked', () => {
  machineRender(
    <>
      <NotificationStatus />
      <OrderStatus />
    </>,
    { spec: mockStore({ state: STATES.DELIVERY }) },
  );
  const button = screen.getByRole('button', { name: 'Cancel Payment' });
  fireEvent.click(button);
  const linkElement = screen.queryByText(/Cancel can be made only on ordering status/i);
  expect(linkElement).toBeInTheDocument();
});

test('test on cancel payment in ordering mode - should be successful', () => {
  machineRender(
    <>
      <NotificationStatus />
      <OrderStatus />
    </>,
    { spec: mockStore({ state: STATES.ORDERING, paid: 10 }) },
  );
  const paidElement = screen.getByText(/You have paid 10 ₪/i);
  expect(paidElement).toBeInTheDocument();
  const button = screen.getByRole('button', { name: 'Cancel Payment' });
  fireEvent.click(button);
  const paidElementAfterCancel = screen.getByText(/You have paid 0 ₪/i);
  expect(paidElementAfterCancel).toBeInTheDocument();
});

test('Order status snapshot', () => {
  const { asFragment } = machineRender(
    <OrderStatus />,
  );
  expect(asFragment()).toMatchSnapshot();
});
