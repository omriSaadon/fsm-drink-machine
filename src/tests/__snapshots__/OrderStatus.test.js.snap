// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Order status snapshot 1`] = `
<DocumentFragment>
  <div
    class="ant-card ant-card-bordered"
  >
    <div
      class="ant-card-head"
    >
      <div
        class="ant-card-head-wrapper"
      >
        <div
          class="ant-card-head-title"
        >
          Order Status
        </div>
      </div>
    </div>
    <div
      class="ant-card-body"
    >
      <div>
        You have paid 0 â‚ª
      </div>
      <button
        class="ant-btn ant-btn-round ant-btn-sm"
        type="button"
      >
        <span>
          Cancel Payment
        </span>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
